README.md
# OpenStorage API Website
This stores the sources and output for https://libopenstorage.github.io. This website is created using
[gitbook](https://github.com/GitbookIO/gitbook-cli). All sources and markdown files are in the `docs` directory. The
output html created by gitbook is located in `w` directory.

## Environment setup
First, you may need to install [npm](https://nodejs.org/en/download/). Then install gitbook:

```
$ npm install -g gitbook-cli
```

To build:

```
make
```

To serve locally on `http://localhost:4000` while editing, run:

```
make serve
```

## Generating API Documentation
To generate it, you will need to install [protoc-gen-doc](https://github.com/lpabon/protoc-gen-doc/releases/tag/v1.1.1)


Then type `make api` to generate the API documentation.

## Releasing
When releasing new version use the following method to maintain that release
version on the documentation website:

1. Copy the docs dir to a new `vX.X.X` dir.
1. Edit the only non-generated link: `vX.X.X/rest-api.md` for the Swagger UI
1. Edit `vX.X.X/swagger-ui/index.html` and adjust the `url:` line to have `/w/vX.X.X/api/api.swagger.json`
1. Add the build to the Makefile `build:` tags
1. Edit `docs/releases.md` accordingly
1. Type `make build`
1. git add the new `vX.X.X` dir and the entire `w/` dir

## More information
* [Gitbook](https://toolchain.gitbook.com/)
* Examples gitbook json setup
    * https://github.com/poppy-project/poppy-docs/blob/master/book.json


docs/releases.md
# Releases

The following table has a set of OpenStorage SDK releases with their respective
documentation and client software.

| Version | Documentation | Clients |
| ------- | ------------- | ------- |
| v0.9.0 | [Reference](https://libopenstorage.github.io/w/v0.9.0) | [Source](https://github.com/libopenstorage/openstorage-sdk-clients/releases/tag/v0.9.0)


examples/python/app.py
import grpc
import jwt
import datetime
import hashlib
from openstorage import api_pb2
from openstorage import api_pb2_grpc


claims = {}
claims['sub'] = 'admin'
claims['iat'] = datetime.datetime.utcnow()
claims['exp'] = datetime.datetime.utcnow() + datetime.timedelta(minutes=10)
claims['name'] = 'Luis Pabon'
claims['email'] = 'luis@portworx.com'
secret='ZDBlMTY3NDhiYTlmZGMwYzQyNTQ2Y2I5Mjg4YzkwNjAgIC0K'
token = jwt.encode(claims, secret, algorithm='HS256')
metadata=[('authorization', 'bearer ' + token)]

channel = grpc.insecure_channel('localhost:9100')

try:
    # Cluster connection
    clusters = api_pb2_grpc.OpenStorageClusterStub(channel)
    ic_resp = clusters.InspectCurrent(api_pb2.SdkClusterInspectCurrentRequest(), metadata=metadata)
    print('Connected to {0} with status {1}'.format(
        ic_resp.cluster.id,
        api_pb2.Status.Name(ic_resp.cluster.status)
    ))

    # Create a volume
    volumes = api_pb2_grpc.OpenStorageVolumeStub(channel)
    v_resp = volumes.Create(api_pb2.SdkVolumeCreateRequest(
        name="myvol",
        spec=api_pb2.VolumeSpec(
            size=100*1024*1024*1024,
            ha_level=3,
        )
    ), metadata=metadata)
    print('Volume id is {0}'.format(v_resp.volume_id))

    # Create a snapshot
    snap = volumes.SnapshotCreate(api_pb2.SdkVolumeSnapshotCreateRequest(
        volume_id=v_resp.volume_id,
        name="mysnap"
    ), metadata=metadata)
    print('Snapshot created with id {0}'.format(snap.snapshot_id))

    # Create a credentials
    creds = api_pb2_grpc.OpenStorageCredentialsStub(channel)
    cred_resp = creds.Create(api_pb2.SdkCredentialCreateRequest(
        name='aws',
        aws_credential=api_pb2.SdkAwsCredentialRequest(
            access_key='dummy',
            secret_key='dummy',
            endpoint='dummy',
            region='dummy'
        )
    ), metadata=metadata)
    print('Credential id is {0}'.format(cred_resp.credential_id))

    # Create backup
    backups = api_pb2_grpc.OpenStorageCloudBackupStub(channel)
    backup_resp = backups.Create(api_pb2.SdkCloudBackupCreateRequest(
        volume_id=v_resp.volume_id,
        credential_id=cred_resp.credential_id,
        full=False
    ), metadata=metadata)

    # Get status
    status_resp = backups.Status(api_pb2.SdkCloudBackupStatusRequest(
        volume_id=v_resp.volume_id
    ), metadata=metadata)
    backup_status = status_resp.statuses[v_resp.volume_id]
    print('Status of the backup is {0}'.format(
        api_pb2.SdkCloudBackupStatusType.Name(backup_status.status)
    ))

    # Show history
    history = backups.History(api_pb2.SdkCloudBackupHistoryRequest(
        src_volume_id=v_resp.volume_id
    ))
    print('Backup history for volume {0}'.format(v_resp.volume_id))
    for item in history.history_list:
        print('Time:{0} Status:{1}'.format(
            item.timestamp.ToJsonString(),
            api_pb2.SdkCloudBackupStatusType.Name(item.status)
        ))

except grpc.RpcError as e:
    print('Failed: code={0} msg={1}'.format(e.code(), e.details()))


examples/golang/main.go
// Example code
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"time"

	jwt "github.com/dgrijalva/jwt-go"

	"github.com/golang/protobuf/ptypes"
	api "github.com/libopenstorage/openstorage-sdk-clients/sdk/golang"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/status"
)

type OpenStorageSdkToken struct{}

func (t OpenStorageSdkToken) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
	secret := "thisistheadminkey"

	// Create Token
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		// Set issuer
		"sub": "admin",
		// Set issued at time
		"iat": time.Now().Unix(),
		// Set expiration
		"exp": time.Now().Add(time.Minute * 5).Unix(),
		// Name
		"name": "Luis Pabon",
		// Email
		"email": "luis@portworx.com",
	})

	// Sign the token
	signedtoken, err := token.SignedString([]byte(secret))
	if err != nil {
		return nil, err
	}

	return map[string]string{
		"authorization": "bearer " + signedtoken,
	}, nil
}

func (t OpenStorageSdkToken) RequireTransportSecurity() bool {
	return false
}

func main() {

	ctx := context.Background()
	token := OpenStorageSdkToken{}

	// Setup connection
	grpccreds, err := credentials.NewClientTLSFromFile("cert.pem", "")
	if err != nil {
		fmt.Printf("Creds failed: %v", err)
		os.Exit(1)
	}

	// Setup a connection
	conn, err := grpc.Dial(":9100", grpc.WithTransportCredentials(grpccreds), grpc.WithPerRPCCredentials(token))
	if err != nil {
		fmt.Printf("Error: %v", err)
		os.Exit(1)
	}

	// Create a cluster client
	cluster := api.NewOpenStorageClusterClient(conn)

	// Print the cluster information
	clusterInfo, err := cluster.InspectCurrent(
		ctx,
		&api.SdkClusterInspectCurrentRequest{})
	if err != nil {
		gerr, _ := status.FromError(err)
		fmt.Printf("A Error Code[%d] Message[%s]\n",
			gerr.Code(), gerr.Message())
		os.Exit(1)
	}
	fmt.Printf("Connected to Cluster %s\n",
		clusterInfo.GetCluster().GetId())
	fmt.Println()

	// Create a 100Gi volume
	volumes := api.NewOpenStorageVolumeClient(conn)
	v, err := volumes.Create(
		ctx,
		&api.SdkVolumeCreateRequest{
			Name: "myvol",
			Spec: &api.VolumeSpec{
				Size:    100 * 1024 * 1024 * 1024,
				HaLevel: 3,
			},
		})
	if err != nil {
		gerr, _ := status.FromError(err)
		fmt.Printf("Error Code[%d] Message[%s]\n",
			gerr.Code(), gerr.Message())
		os.Exit(1)
	}
	fmt.Printf("Volume 100Gi created with id %s\n", v.GetVolumeId())
	fmt.Println()

	// Take a snapshot
	snap, err := volumes.SnapshotCreate(
		ctx,
		&api.SdkVolumeSnapshotCreateRequest{
			VolumeId: v.GetVolumeId(),
			Name:     fmt.Sprintf("snap-%v", time.Now().Unix()),
		},
	)
	if err != nil {
		gerr, _ := status.FromError(err)
		fmt.Printf("Error Code[%d] Message[%s]\n",
			gerr.Code(), gerr.Message())
		os.Exit(1)
	}
	fmt.Printf("Snapshot with id %s was create for volume %s\n",
		snap.GetSnapshotId(),
		v.GetVolumeId())
	fmt.Println()

	// Create mock credentials
	creds := api.NewOpenStorageCredentialsClient(conn)
	credResponse, err := creds.Create(context.Background(),
		&api.SdkCredentialCreateRequest{
			Name: "aws",
			CredentialType: &api.SdkCredentialCreateRequest_AwsCredential{
				AwsCredential: &api.SdkAwsCredentialRequest{
					AccessKey: "dummy-access",
					SecretKey: "dummy-secret",
					Endpoint:  "dummy-endpoint",
					Region:    "dummy-region",
				},
			},
		})
	if err != nil {
		gerr, _ := status.FromError(err)
		fmt.Printf("Error Code[%d] Message[%s]\n",
			gerr.Code(), gerr.Message())
		os.Exit(1)
	}
	credID := credResponse.GetCredentialId()
	fmt.Printf("Credentials created with id %s\n", credID)
	fmt.Println()

	// Create a backup to a cloud provider of our volume
	cloudbackups := api.NewOpenStorageCloudBackupClient(conn)
	_, err = cloudbackups.Create(context.Background(),
		&api.SdkCloudBackupCreateRequest{
			VolumeId:     v.GetVolumeId(),
			CredentialId: credID,
		})
	if err != nil {
		gerr, _ := status.FromError(err)
		fmt.Printf("Error Code[%d] Message[%s]\n",
			gerr.Code(), gerr.Message())
		os.Exit(1)
	}
	fmt.Printf("Backup started for volume %s\n", v.GetVolumeId())

	// Now check the status of the backup
	backupStatus, err := cloudbackups.Status(context.Background(),
		&api.SdkCloudBackupStatusRequest{
			VolumeId: v.GetVolumeId(),
		})
	if err != nil {
		gerr, _ := status.FromError(err)
		fmt.Printf("Error Code[%d] Message[%s]\n",
			gerr.Code(), gerr.Message())
		os.Exit(1)
	}
	for volID, status := range backupStatus.GetStatuses() {
		// There will be only one value in the map, but we use
		// a for-loop as an example.
		b, _ := json.MarshalIndent(status, "", "  ")
		fmt.Printf("Backup status for volume: %s\n"+
			"Type: %s\n"+
			"Status: %s\n"+
			"Full JSON Response: %s\n",
			volID,
			status.GetOptype().String(),
			status.GetStatus().String(),
			string(b))
	}
	fmt.Println()

	// Backup History
	historyResp, err := cloudbackups.History(context.Background(),
		&api.SdkCloudBackupHistoryRequest{
			SrcVolumeId: v.GetVolumeId(),
		})
	if err != nil {
		gerr, _ := status.FromError(err)
		fmt.Printf("Error Code[%d] Message[%s]\n",
			gerr.Code(), gerr.Message())
		os.Exit(1)
	}

	fmt.Printf("Backup history for volume %s:\n", v.GetVolumeId())
	for _, history := range historyResp.GetHistoryList() {

		timestamp, _ := ptypes.Timestamp(history.GetTimestamp())
		fmt.Printf("Volume:%s \tttime:%v \tstatus:%v\n",
			history.GetSrcVolumeId(),
			timestamp,
			history.GetStatus())
	}
	fmt.Println()
}
